library/std/src/thread/mod.rs

```diff
@@ -887,6 +879,22 @@ pub fn sleep(dur: Duration) {
 ///
 /// * It can be implemented very efficiently on many platforms.
 ///
+/// # Memory Orderings
+///
+/// Calls to `park` _synchronize-with_ calls to `unpark`, meaning that memory
+/// operations performed before a call to `unpark` are made visible to the thread that
+/// consumes the token and returns from `park`. Note that all `park` and `unpark`
+/// operations for a given thread form a total order and `park` synchronizes-with
+/// _all_ prior `unpark` operations.
+///
+/// In atomic ordering terms, `unpark` performs a `Release` operation and `park`
+/// performs the corresponding `Acquire` operation. Calls to `unpark` for the same
+/// thread form a [release sequence].
+///
+/// Note that being unblocked does not imply a call was made to `unpark`, because
+/// wakeups can also be spurious. For example, a valid, but inefficient,
+/// implementation could have `park` and `unpark` return immediately without doing anything.
```

Comment:
That last sentence sounds wrong now? That implementation would not establish the described synchronization edges.

Reply:
Don't we want to guarantee that eventually, there'll be a non-spurious wakeup (if `unpark` was previously called in another thread)?

Reply:
I'm not sure what such a guarantee would help with. We also don't guarantee that park() will wake up within a second (or an hour or a year) after unpark() or something like that..

When considering the soundness of code using park() and unpark(), it's a useful limitus test to think what would happen if both park() and unpark() did exactly nothing. If that would result in unsoundness, the code is unsound. So that's a useful thing to think about. But I don't see what a guarantee of 'park() will consume the token eventually' would really help with.

Reply:
I was thinking we want to provide some kind of liveness guarantee ("forward progress" in C++ terms). But I guess we don't make a good effort of documenting our liveness guarantees for any part of the concurrency stack so no reason to start here and now.

Also before we add anything like it I should come up with an example where that liveness guarantee is truly needed. Currently we have "either the wakeup was spurious and so there'll be another wakeup (as the token was not consumed), or it was proper and there was synchronization", which (together with fairness of memory event propagation) should be sufficient.